### PR TITLE
Remote cluster management

### PR DESCRIPTION
**Pull request by TerryHowe**
remotely manager clusters

### CODE REVIEW

File: pkg/packages/manager.go [hunk 1]
```diff
@@ -132,6 +148,11 @@ func processInstalled(mc *ManagerContext) bool {
 }
 
 func processUninstalling(mc *ManagerContext) bool {
+	if err := mc.PackageDriver.Initialize(mc.Ctx, mc.getClusterName()); err != nil {
+		mc.Package.Status.Detail = err.Error()
+		mc.Log.Error(err, "Initialization failed")
+		return true
```
[the remark this answers is not in the document]
Comment (TerryHowe) [reply pkg/packages/manager.go:154]: We aren't going to attempt to delete packages if someone deletes a cluster. If they want to delete the packages first for some reason, they should do that before the cluster. I can't think of any need for that at this time unless a package delete would return some resource and that just isn't the case right now.

File: pkg/driver/helmdriver.go [hunk 2]
```diff
@@ -26,44 +26,49 @@ const (
...
 }
 
 var _ PackageDriver = (*helmDriver)(nil)
 
-func NewHelm(log logr.Logger, secretAuth auth.Authenticator) (*helmDriver, error) {
-	settings := cli.New()
+func NewHelm(log logr.Logger, secretAuth auth.Authenticator, kubeconfigClient auth.KubeconfigClient) (*helmDriver, error) {
+	return &helmDriver{
+		secretAuth:       secretAuth,
+		kubeconfigClient: kubeconfigClient,
```
Comment (jonahjon) [inline pkg/driver/helmdriver.go:41]: Is the `kubeconfigClient` a required field for the NewHelm function now?

Comment (TerryHowe) [reply pkg/driver/helmdriver.go:41]: yes

File: pkg/packages/manager.go [hunk 3]
```diff
@@ -19,6 +20,7 @@ const (
 	retryLong      = time.Duration(60) * time.Second
 	retryVeryLong  = time.Duration(180) * time.Second
 	sourceRegistry = "sourceRegistry"
+	prefix         = api.PackageNamespace + "-"
```
Comment (jonahjon) [inline pkg/packages/manager.go:23]: Is this needed?

Comment (TerryHowe) [reply pkg/packages/manager.go:23]: it gets used 3 places 🤷

File: pkg/packages/manager.go [hunk 4]
```diff
@@ -32,11 +34,19 @@ type ManagerContext struct {
...
 }
 
-func (mc *ManagerContext) SetUninstalling(name string) {
+func (mc *ManagerContext) SetUninstalling(namespace string, name string) {
+	mc.Package.Namespace = namespace
 	mc.Package.Name = name
 	mc.Package.Status.State = api.StateUninstalling
 }
 
+func (mc *ManagerContext) getClusterName() string {
```
Comment (jonahjon) [inline pkg/packages/manager.go:43]: Is this just used for namespaces with a trailing `-` in the namepace, and is that common?

File: pkg/packages/manager.go [hunk 5]
```diff
@@ -98,40 +107,43 @@ func processInstalled(mc *ManagerContext) bool {
...
+		return true
 	}
 	mc.RequeueAfter = retryVeryLong
 
 	return false
 }
 
 func processUninstalling(mc *ManagerContext) bool {
+	if err := mc.PackageDriver.Initialize(mc.Ctx, mc.getClusterName()); err != nil {
+		mc.Package.Status.Detail = err.Error()
```
Comment (jonahjon) [inline pkg/packages/manager.go:143]: Does this set the status to error, and is that intended?

File: pkg/driver/helmdriver.go (hunk 2 above)
Comment (jonahjon) [reply pkg/driver/helmdriver.go:41]: Ok I'll need some changes in generatebundle, since we use it there too.

File: pkg/packages/manager.go (hunk 3 above)
Comment (jonahjon) [reply pkg/packages/manager.go:23]: I mean why/where are namespaces being created with `-` in them?

Comment (TerryHowe) [reply pkg/packages/manager.go:23]: oh, that is just the namespace prefix. We will be creating these namespaces for each pbc. If there is a pbc for the billy cluster, there will be a namespace `eksa-packages-billy` for that cluster where packages CRs should go

File: pkg/packages/manager.go (hunk 4 above)
Comment (TerryHowe) [reply pkg/packages/manager.go:43]: see the other comment

File: pkg/packages/manager.go (hunk 5 above)
Comment (TerryHowe) [reply pkg/packages/manager.go:143]: yes, this allows the user to see the status of an install without looking on controller logs

Comment (TerryHowe) [reply pkg/packages/manager.go:143]: For example, the "Packages namespace..." is the detail here:
```
default                package.packages.eks.amazonaws.com/my-hello   hello-eks-anywhere   119m    unknown                                                      0.1.1-bc7dc6bb874632972cd92a2bca429a846f7aa785 (latest)   Packages namespaces must start with: eksa-packages
````

Comment (jonahjon) [reply pkg/packages/manager.go:149]: O ok, and this is on the uninstalls

File: pkg/packages/manager.go (hunk 3 above)
Comment (jonahjon) [reply pkg/packages/manager.go:23]: ooo ok, yea I think that's fine then.

File: pkg/authenticator/kubeconfig.go [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+	}
+
+	secretName := clusterName + "-kubeconfig"
+	nn := types.NamespacedName{
+		Namespace: eksAnywhereNamespace,
+		Name:      secretName,
+	}
+	var kubeconfigSecret corev1.Secret
+	if err = kc.Client.Get(ctx, nn, &kubeconfigSecret); err != nil {
+		return "", fmt.Errorf("getting kubeconfig secret: %v", err)
```
[the remark this answers is not in the document]
Comment (TerryHowe) [reply pkg/authenticator/kubeconfig.go:42]: https://github.com/aws/eks-anywhere-packages/pull/480

File: pkg/authenticator/kubeconfig.go [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+
+func NewKubeconfigClient(client client.Client) *kubeconfigClient {
+	return &(kubeconfigClient{Client: client})
+}
+
+var _ KubeconfigClient = (*kubeconfigClient)(nil)
+
+func (kc *kubeconfigClient) GetKubeconfig(ctx context.Context, clusterName string) (fileName string, err error) {
+	if clusterName == "" {
+		return "", nil
```
[the remark this answers is not in the document]
Comment (TerryHowe) [reply pkg/authenticator/kubeconfig.go:32]: https://github.com/aws/eks-anywhere-packages/pull/480

File: pkg/authenticator/kubeconfig_test.go [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+		nn := types.NamespacedName{
+			Namespace: "eksa-system",
+			Name:      "billy-kubeconfig",
+		}
+		mockClient.EXPECT().Get(ctx, nn, gomock.Any()).DoAndReturn(setKubeConfigSecret(&kubeconfigSecret))
+
+		fileName, err := sut.GetKubeconfig(ctx, "billy")
+		assert.NoError(t, err)
+		assert.Equal(t, "billy-kubeconfig", fileName)
+		os.Remove(fileName)
```
[the remark this answers is not in the document]
Comment (TerryHowe) [reply pkg/authenticator/kubeconfig_test.go:38]: https://github.com/aws/eks-anywhere-packages/pull/482

File: pkg/authenticator/kubeconfig.go [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+	nn := types.NamespacedName{
+		Namespace: eksAnywhereNamespace,
+		Name:      secretName,
+	}
+	var kubeconfigSecret corev1.Secret
+	if err = kc.Client.Get(ctx, nn, &kubeconfigSecret); err != nil {
+		return "", fmt.Errorf("getting kubeconfig secret: %v", err)
+	}
+
+	err = os.WriteFile(secretName, kubeconfigSecret.Data["value"], 0600)
```
[the remark this answers is not in the document]
Comment (TerryHowe) [reply pkg/authenticator/kubeconfig.go:45]: https://github.com/aws/eks-anywhere-packages/pull/482

**Pull request merged**